### PR TITLE
Validate/Generate the returned state value.

### PR DESCRIPTION
**Pull request by guest20**
By making this a separate parameter to get_token, folks already passing/validating state can keep their code how it is.

### CODE REVIEW

**Comment by guest20** [comment]
Closes #57

File: lib/Mojolicious/Plugin/OAuth2.pm [hunk 1]
```diff
@@ -3,6 +3,7 @@ use Mojo::Base 'Mojolicious::Plugin';
 
 use Mojo::Promise;
 use Mojo::UserAgent;
+use Mojo::Util qw[ sha1_sum ];
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OAuth2.pm:6]: This should be `qw(sha1_sum)` instead to follow the Mojolicious standards.

File: lib/Mojolicious/Plugin/OAuth2.pm [hunk 2]
```diff
@@ -96,10 +103,19 @@ sub _get_authorize_url {
   $authorize_url->query->append(client_id => $provider_args->{key}, redirect_uri => $args->{redirect_uri});
   $authorize_url->query->append(scope => $args->{scope}) if defined $args->{scope};
   $authorize_url->query->append(state => $args->{state}) if defined $args->{state};
+  $authorize_url->query->append(state => $c->session->{$args->{state_session_key}} //= $self->_make_state)
+    if defined $args->{state_session_key};
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OAuth2.pm:107]: I think you want to use `flash` instead of `session` here.

https://docs.mojolicious.org/Mojolicious/Plugin/DefaultHelpers#flash

File: lib/Mojolicious/Plugin/OAuth2.pm [hunk 3]
```diff
@@ -96,10 +103,19 @@ sub _get_authorize_url {
...
   $authorize_url->query($args->{authorize_query}) if exists $args->{authorize_query};
   $authorize_url;
 }
 
+my $start = time;
+sub _make_state {
+  my ($self) = @_;
+  # Hopefully nobody will be able to guess this value: 
+  sha1_sum join '-', $start, time, $$, rand
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OAuth2.pm:116]: To make it a bit more secure, you should use `Time::HiRes::time()`.
`$start` can be replaced with `$^T`.
I also like mixing in `$$` and `$<`.
Last, this line lacks semicolon.

File: lib/Mojolicious/Plugin/OAuth2.pm [hunk 4]
```diff
@@ -112,10 +128,31 @@ sub _get_token {
 
   # No error or code response from provider callback URL
   unless ($c->param('code')) {
-    $c->redirect_to($self->_get_authorize_url($c, $args)) if $args->{redirect} // 1;
+    my $url = $self->_get_authorize_url($c, $args);
+    $c->app->log->debug("no code, redirecting to $url");
+    $c->redirect_to( $url) if $args->{redirect} // 1;
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OAuth2.pm:133]: This whole block seems unrelated to the PR. Please explain how it's related, if it is.

File: lib/Mojolicious/Plugin/OAuth2.pm [hunk 5]
```diff
@@ -112,10 +128,31 @@ sub _get_token {
...
+    my $param_state = $c->param('state');
+
+    die "state missing"  if not defined $param_state;
+    if ($c->session->{$args->{state_session_key}} ne $param_state) {
+        $c->app->log->error("oauth state check: " . sprintf "state mismatch session(%s) != param(%s)", $session_state, $param_state);
+        die "oauth state mismatch"
+    }
+
+    # We don't want to re-use this value, so:
+    delete $c->session->{$args->{state_session_key}};
```
Comment (jhthorsen) [inline lib/Mojolicious/Plugin/OAuth2.pm:148]: You don't need this if you use `flash`.

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 2 above)
Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:107]: How do you `//=` a flash key?

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 4 above)
Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:133]: Personally, I think of a debug logging line as a comment, but one that also gets a chance to be helpful while the program is running.

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 3 above)
Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:116]: `Time::HiRes` won't make `$^T` `HiRes`, so those two points don't really mesh very well.

`$$` is already there, adding`$<` is no problem.

Comment (jhthorsen) [reply lib/Mojolicious/Plugin/OAuth2.pm:116]: I mean: Replace `time, $$, rand` with `Time::HiRes::time, $$, rand`.

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 4 above)
Comment (jhthorsen) [reply lib/Mojolicious/Plugin/OAuth2.pm:133]: What about the rest of the code?

Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:133]: I put the url we're about to redirect to into `$url`, so as to not call `_get_authorize_url` twice.

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 3 above)
Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:116]: What do you mean about `$^T`? Should that be involved or not?

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 4 above)
Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:133]: Oh, maybe you mean "what about the logging in the rest of the code"?

I personally want to keep an eye on the attempted/failed/error'ed oauth logins in my app. Just so i can check the logs and find out of some bot has started spamming attempts against my oauth endpoint(s).

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 3 above)
Comment (jhthorsen) [reply lib/Mojolicious/Plugin/OAuth2.pm:116]: >  $start can be replaced with $^T.
Yes, you don't need $start.

Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:116]: even though `$^T` is not in fractional seconds? Maybe it's worth taking the "a bit more secure" approach and making `$start` a `Time::HiRes::time()` too?

**Comment by guest20** [comment]
I think this PR could be lot cleaner if the code wasn't guarded by `$args->{state_session_key}`, to remove it I'd need:
  - to use the provider name as a key in the flash: `$c->flash("code_for_$provider" =>... ->_make_state)`
  - provide an option like  `$self->providers->{$provider}{ignore_returned_state} = 1` to turn off `state` validation

Is it worth making these changes to the PR or would enabling state validation by default be too disruptive?

File: lib/Mojolicious/Plugin/OAuth2.pm (hunk 4 above)
Comment (guest20) [reply lib/Mojolicious/Plugin/OAuth2.pm:139]: This logging is now guarded by $ENABLE_LOGGING, which defaults to off.

Maybe someone can have a look at the tests and see if they cover enough?

**Comment by jhthorsen** [comment]
I'm going to close this, since most of the comments have not been addressed.
